From f6f3eb68cd4f735c4c49e612e8398afe5892c482 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Sun, 7 Sep 2014 17:16:08 -0400
Subject: [pci] hotplug: Prevent NULL dereference during pciehp probe

Message-id: <20140907171608.4269.3106.stgit@gir.stowe>
Patchwork-id: 91625
O-Subject: [RHEL7.1 PATCH] PCI: Prevent NULL dereference during pciehp probe
Bugzilla: 1133107
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1133107
Upstream Status: 62e4492c3063048a163d238cd1734273f2fc757d
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=7925983
Tested: IBM tested this patch on top of a kernel-3.10.0-123 kernel.

commit 62e4492c3063048a163d238cd1734273f2fc757d
Author: Andreas Noever <andreas.noever@gmail.com>
Date:   Mon Jun 9 23:03:32 2014 +0200

    PCI: Prevent NULL dereference during pciehp probe

    pciehp assumes that dev->subordinate exists.  But we do not assign a bus if
    we run out of bus numbers during enumeration.  This leads to a NULL
    dereference in init_slot() (and other places).

    Change pciehp_probe() to return -ENODEV when no subordinate bus is present.

    Signed-off-by: Andreas Noever <andreas.noever@gmail.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/hotplug/pciehp_core.c b/drivers/pci/hotplug/pciehp_core.c
index 0e0a2ff..e5abcaf 100644
--- a/drivers/pci/hotplug/pciehp_core.c
+++ b/drivers/pci/hotplug/pciehp_core.c
@@ -255,6 +255,13 @@ static int pciehp_probe(struct pcie_device *dev)
  else if (pciehp_acpi_slot_detection_check(dev->port))
   goto err_out_none;
 
+ if (!dev->port->subordinate) {
+  /* Can happen if we run out of bus numbers during probe */
+  dev_err(&dev->device,
+   "Hotplug bridge without secondary bus, ignoring\n");
+  goto err_out_none;
+ }
+
  ctrl = pcie_init(dev);
  if (!ctrl) {
   dev_err(&dev->device, "Controller initialization failed\n");
-- 
1.7.1